### ISSUE TITLE
Encrypt the private key file on disk

### ISSUE THREAD

**Issue by yutianwu** [tendermint/tendermint#2657]
So far, we store private key file on disk unencrypted. 

But if we use the unencrypted private key file in production, we will be at the risk of leaking the private key which is classified. For example, people who maintain the validator will have access to the machine host the validator.

So can we encrypt the private key file on disk? just like keystore of Ethereum.

**Comment by yutianwu** [comment]
hi, @ebuchman, if you agree the issue, I will work on this. 

IMHO, I want to encrypt the `PrivKey` field in `FilePV` in a new field named `EncryptedPrivKey`.
``` go
type FilePV struct {
	Address       types.Address  `json:"address"`
	PubKey        crypto.PubKey  `json:"pub_key"`
	LastHeight    int64          `json:"last_height"`
	LastRound     int            `json:"last_round"`
	LastStep      int8           `json:"last_step"`
	LastSignature []byte         `json:"last_signature,omitempty"` // so we dont lose signatures XXX Why would we lose signatures?
	LastSignBytes cmn.HexBytes   `json:"last_signbytes,omitempty"` // so we dont lose signatures XXX Why would we lose signatures?
	PrivKey       crypto.PrivKey `json:"-"`

	EncryptedPrivKey EncryptedKey `json:"encrypted_priv_key"`

	// For persistence.
	// Overloaded for testing.
	filePath string
	mtx      sync.Mutex
}
```

So when we want to generate a private key file or want to load the private key file, we need a auth to decrypt it. And `PrivKey` only exists in memory and will not exported to file. The format of private key file should be like this(I just refer to the ethereum key store file):
```json
{
  "address": "479928A17D076ACD833A41D7F8B58B6AF6B53AE5",
  "pub_key": {
    "type": "tendermint/PubKeyEd25519",
    "value": "Hgfsm2EX1dU44NjvRq9FgcTXaErdgOXaFut4mG5cKJw="
  },
  "last_height": "0",
  "last_round": "0",
  "last_step": 0,
  "encrypted_priv_key": {
    "crypto": {
      "cipher": "aes-128-ctr",
      "ciphertext": "0e6f91296c8867e722878d64e19c89c49a48ab70a81afd354e7358adb3bd8679d2cb9c96ea177c73f1c6b94ae7c883fead7d49b23c8c69b91caa15ea8fe5a9b91eed3f0f6b4d",
      "cipherparams": {
        "iv": "349b25d741b4cd28b34f54264e550d71"
      },
      "kdf": "scrypt",
      "kdfparams": {
        "n": "8",
        "r": "8",
        "p": "8",
        "dklen": "12",
        "salt": "c97b60e59f0fd9d7fa9b6e6f4aaf52ad6ce9941de8a7c8f96abff0be183a4762"
      },
      "mac": "3cec1fdfe2b4638a74c1a5969d4aa8a167b2ad3bf2a1d551ffb2be3354e87633"
    },
    "version": "1"
  }
}
```

What's your opinion?

**Comment by ValarDragon** [comment]
I think the way we want to do this is what s described here: https://github.com/tendermint/tendermint/issues/1181

Basically we separate the mutable part and immutable part of this. Then the immutable part can be encrypted on disk.

**Comment by yutianwu** [comment]
@ValarDragon I see. Actually I think it is good to split mutable section and immutable section, that why I add a `PrivKey ` field. 

So we want to files:
1. one contains immutable fields like public key, private key:
```json
{
  "pub_key": {
    "type": "tendermint/PubKeyEd25519",
    "value": "dGxtDK0uYz1y7ijeNDk9NjjR8q54EOF3+kk4/9g7KR4="
  },
  "encrypted_priv_key": {
    "crypto": {
      "cipher": "aes-128-ctr",
      "ciphertext": "0e6f91296c8867e722878d64e19c89c49a48ab70a81afd354e7358adb3bd8679d2cb9c96ea177c73f1c6b94ae7c883fead7d49b23c8c69b91caa15ea8fe5a9b91eed3f0f6b4d",
      "cipherparams": {
        "iv": "349b25d741b4cd28b34f54264e550d71"
      },
      "kdf": "scrypt",
      "kdfparams": {
        "n": "8",
        "r": "8",
        "p": "8",
        "dklen": "12",
        "salt": "c97b60e59f0fd9d7fa9b6e6f4aaf52ad6ce9941de8a7c8f96abff0be183a4762"
      },
      "mac": "3cec1fdfe2b4638a74c1a5969d4aa8a167b2ad3bf2a1d551ffb2be3354e87633"
    },
    "version": "1"
  }
}
```
2. one contains mutable fields like height, round, step and etc.
```json
{
  "address": "479928A17D076ACD833A41D7F8B58B6AF6B53AE5",
  "last_height": "0",
  "last_round": "0",
  "last_step": 0,
  "last_signature": {
    "type": "tendermint/SignatureEd25519",
    "value": "4T8AE7Ab4/zPfEp0ltYWNziC/QVp6ccImlIuKUdhxEJpYBSdX1Ah458GEcB5eDnY5t+7bkvjmIHw1v63hH4VBg=="
  }
}
```

So can we talk about what else do you want to add to the immutable section, so I can start work on this. thx.

**Comment by ebuchman** [comment]
@yutianwu , want to start with a PR implementing #1181, and then we can tackle the encryption?

**Comment by ebuchman** [comment]
So in [ADR-008](https://github.com/tendermint/tendermint/blob/v0.25.0/docs/architecture/adr-008-priv-validator.md), we decided to not support encryption for keys in Tendermint itself, but to use an external key signer instead, ie. the [KMS](https://github.com/tendermint/kms). Have you looked at using that?


That said, we should probably support generating encrypted node and validator keys in Tendermint itself, but have an unsafe mode for development/testing where it remains plaintext.

We have some key encryption code in the SDK repo, for on-disk user keys - https://github.com/cosmos/cosmos-sdk/blob/develop/crypto/keys/mintkey/mintkey.go, but it uses gpg-like armor instead of the ETH format. In any case, we should probably be consistent.

**Comment by yutianwu** [comment]
Of course

**Comment by ebuchman** [comment]
1. We want to split the mutable/immutable regardless, but for encrypting the immutable, I think we want an external FilePV process
2. Thanks for pointing this out. We should ideally be using an authenticated encryption cipher like xsalsa20poly1305 or aes-gcm. It looks like we're using xsalsa20 alone, without the integrity checks. @ValarDragon do you know what happened here? Were we really never authenticating?

**Comment by ValarDragon** [comment]
We are using `xsalsa20poly1305`. This is in https://github.com/golang/crypto/blob/master/nacl/secretbox/secretbox.go#L76. Perhaps the module name should be named "xsalsa20poly1305symmetric", instead of omitting the poly1305.

**Comment by ebuchman** [comment]
Ah of course. Thank you Dev! All hail NaCl

**Comment by ValarDragon** [comment]
I think I ended up calling it symmetric just because that was the interface we were making it abide. Naming it `xsalsa20poly1305` implied it was AEAD, which was not the API we provided. 

I'm fairly confident I wrote a generic converter from AEAD to object abiding our symmetric interface, but I can't find it. I think eventually we should just re-introduce that, and then make xsalsa20poly1305 the default in there. (I'm puzzled as to where this code went though, I really thought I made a package in the crypto folder called symmetric)
